/* Copyright 2016 Jiang Chen <criver@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = 'proto3';

package gbdt;

message CategoricalSplit {
  repeated string category = 1;
  // Internal usage.
  repeated int32 internal_categorical_index = 2;
}

// The split is feature < threshold.
message FloatSplit {
  float threshold = 1;
  // Whether or not the missing data is split to the right child.
  bool missing_to_right_child = 2;
}

message Split {
  string feature = 1;
  FloatSplit float_split = 2;
  CategoricalSplit cat_split = 3;

  double gain = 5;
}

message TreeNode {
  float score = 1;
  Split split = 2;
  TreeNode left_child = 3;
  TreeNode right_child = 4;
}

message Forest {
  repeated TreeNode tree = 1;
  string meta_info = 2;
}